### PR TITLE
Qwen3-Next-MTP Impl

### PR DESCRIPTION
**Pull request by ganyi1996ppo**
## Motivation
This PR enable qwen3-next-mtp in atom
### Launching Script 
```bash
python3 -m atom.entrypoints.openai_server \
  --model $MODEL \
  --gpu-memory-utilization 0.8 \
  --level 1 \
  -tp 4 \
  --server-port 8000 \
  --method "mtp" \
  --num-speculative-tokens 1 \
```

## Test Plan
gsm8k on LMEval

## Test Result
cudagraph:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8415|±  |0.0101|
|     |       |strict-match    |     5|exact_match|↑  |0.8287|±  |0.0104|
```
Acceptance rate: 92.87%

eager:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8650|±  |0.0094|
|     |       |strict-match    |     5|exact_match|↑  |0.8415|±  |0.0101|
```
Acceptance rate: 93.16%

We notice there are some accuracy gap between graph mode and eager mode, and have identify which is caused mainly by executing drafter. It is found that the graph mode on qwen3-next-mtp can also get to 0.86 if we skip drafter directly during inference. We might fix this in the future.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: atom/model_engine/model_runner.py [hunk 1]
```diff
@@ -1359,6 +1383,8 @@ def postprocess(
                 target_logits,
                 bonus_token_ids,
             )
+            # print("sampled tokens: ", sampled_tokens, flush=True)
```
Comment (valarLip) [inline atom/model_engine/model_runner.py]: ..

File: atom/config.py [hunk 2]
```diff
@@ -541,6 +543,18 @@ def hf_config_override(hf_config: PretrainedConfig) -> PretrainedConfig:
...
+            if n_predict != 1:
+                logger.warning(
+                    f"Overriding num_nextn_predict_layers from {n_predict} to 1 "
+                    "(MTP typically uses 1 layer that gets reused)"
+                )
+                n_predict = 1
+            hf_config.update(
+                {"n_predict": n_predict, "architectures": ["Qwen3NextMTPModel"]}
+            )
+        print("hf config is: ", hf_config, flush=True)
```
Comment (valarLip) [inline atom/config.py]: switch to logger.info

File: atom/model_engine/model_runner.py [hunk 3]
```diff
@@ -91,6 +90,9 @@ def __init__(
         self.default_num_rejected_tokens = torch.zeros(
             max_num_batched_tokens, dtype=torch.int32, device=device
         )
+        self.prev_prefills = torch.zeros(
```
Comment (valarLip) [inline atom/model_engine/model_runner.py]: useless?

Comment (ganyi1996ppo) [reply atom/model_engine/model_runner.py]: Indeed, removed

File: atom/model_engine/model_runner.py [hunk 4]
```diff
@@ -131,19 +133,32 @@ def recv_async_output_draft(self) -> np.ndarray:
...
+    def recv_mtp_status_async(self) -> tuple[Optional[np.ndarray]]:
         if not self.rejected_tokens_cpu:
-            return None
+            return None, None
         self.async_copy_event.synchronize()
-        return self.rejected_tokens_cpu.pop(0).numpy()
+        return (
+            self.rejected_tokens_cpu.pop(0).numpy(),
+            self.bonus_tokens_cpu.pop(0).numpy(),
+        )
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/model_engine/model_runner.py]: Done

File: atom/model_engine/model_runner.py [hunk 5]
```diff
@@ -1008,7 +1032,7 @@ def allocate_kv_cache(self, num_kvcache_blocks):
                 hf_config.linear_key_head_dim,
                 hf_config.linear_key_head_dim,
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/model_engine/model_runner.py]: Done

File: atom/model_engine/model_runner.py (hunk 1 above)
Comment (ganyi1996ppo) [reply atom/model_engine/model_runner.py]: Removed

File: atom/model_engine/model_runner.py [hunk 6]
```diff
@@ -1359,6 +1383,8 @@ def postprocess(
                 target_logits,
                 bonus_token_ids,
             )
+            # print("sampled tokens: ", sampled_tokens, flush=True)
+            # print("num bonus tokens: ", num_bonus_tokens, flush=True)
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/model_engine/model_runner.py]: Removed

File: atom/model_ops/attention_mha.py [hunk 7]
```diff
@@ -202,9 +202,13 @@ def paged_attention_triton(
         attn_metadata = fwd_ctx.attn_metadata
 
         o = torch.empty_like(q)
-        num_seqs, num_q_heads_total, head_size = q.shape
+        num_seqs = attn_metadata.context_lens.shape[0]
+        _, num_q_heads_total, head_size = q.shape
         num_blocks, num_kv_heads, _, block_size, _ = k_cache.shape
-        query_group_size = num_q_heads_total // num_kv_heads
+        # assuem all query have same length
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/model_ops/attention_mha.py]: Done

File: atom/config.py [hunk 8]
```diff
@@ -541,6 +543,18 @@ def hf_config_override(hf_config: PretrainedConfig) -> PretrainedConfig:
...
+        if hf_config.model_type == "qwen3_next_mtp":
+            n_predict = getattr(hf_config, "num_nextn_predict_layers", 1)
+            if n_predict != 1:
+                logger.warning(
+                    f"Overriding num_nextn_predict_layers from {n_predict} to 1 "
+                    "(MTP typically uses 1 layer that gets reused)"
+                )
+                n_predict = 1
+            hf_config.update(
+                {"n_predict": n_predict, "architectures": ["Qwen3NextMTPModel"]}
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/config.py:555]: Qwen3-Next don't need this

File: atom/config.py (hunk 2 above)
Comment (ganyi1996ppo) [reply atom/config.py]: sure

File: atom/model_ops/attentions/gdn_attn.py [hunk 9]
```diff
@@ -125,10 +130,33 @@ def __init__(
         }
         self.model_runner.forward_vars.update(gdn_metadata)
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/model_ops/attentions/gdn_attn.py:131]: It doesn't need it

File: atom/models/qwen3_next.py [hunk 10]
```diff
@@ -55,6 +55,64 @@
...
+                boundary : (boundary + take), ...  # type: ignore[misc]
+            ] = loaded_weight[
+                loaded_start_idx : (loaded_start_idx + take)  # type: ignore[misc]
+            ]  # type: ignore[misc]
+
+            # move indexing boundaries
+            boundary += shard_size
+            loaded_boundary += full_dim - extra
+
+    return loader
```
[the remark this answers is not in the document]
Comment (ganyi1996ppo) [reply atom/models/qwen3_next.py:109]: No duplicated code
